### ISSUE TITLE
ENH: Speed up computation of Bessel functions

### ISSUE THREAD

**Issue by LemonBoy** [scipy/scipy#19079]
### Is your feature request related to a problem? Please describe.

I am currently writing some code whose hot loop involves the repeated computation of Bessel's `jn` and `yn` functions for a few orders (`0..50`) and for a large number of points (~200k). Currently SciPy's functions (both in Python and Cython form) are unusably slow, the following one-liner takes tens of seconds on my desktop:

```python
timeit.timeit(lambda: jn([[x] for x in np.arange(51, dtype=int)], np.arange(1e-3, 200, 1e-3)))
```

while MATLAB's counterpart executes in around four seconds, giving an order of magnitude (and a half) between the two languages:

```MATLAB
clear all
x = 1e-3:1e-3:200;
J = zeros(50,length(x));
tic
for i = 0:50-1
    J(i+1,:) = besselj(i, x);
end
toc
```

From what I can gather both Matlab and SciPy use the AMOS package under the hood (with some cases falling back to Cephes) and so I'd expect the performance to be comparable.

How is the vectorization implemented? If it is expressed in terms of `np.vectorize` that'd partially explain the slowdown, from what I see AMOS's functions support computing more than one order/point at the same time and it'd be nice to exploit that interface to save some time.

I'm no NumPy/SciPy wizard so if you have a better idea of what's wrong (or what could be done) I'm all ears.

**Comment by ev-br** [comment]
Since you mentioned Cython, you can fairly easily throw some OpenMP

```
In [60]: a = np.arange(1e-2, 1, 1e-2)

In [61]: x = np.arange(51, dtype=int)

In [62]: %%cython -c=-fopenmp --link-args=-fopenmp
    ...: cimport cython
    ...: from scipy.special.cython_special cimport jv
    ...: from cython.parallel import prange
    ...: import numpy as np
    ...: 
    ...: @cython.boundscheck(False)
    ...: @cython.wraparound(False)
    ...: def jvv(long[::1] x, double[::1] a):
    ...:      cdef double[:, ::1] out = np.empty((x.shape[0], a.shape[0]))
    ...:      cdef int i, j
    ...:      with nogil:
    ...:          for i in prange(x.shape[0]):
    ...:              for j in range(a.shape[0]):
    ...:                 out[i, j] = jv(x[i], a[j])
    ...:      return np.asarray(out)
    ...: 

In [63]: %timeit jvv(x, a)
255 µs ± 5.18 µs per loop (mean ± std. dev. of 7 runs, 1000 loops each)

In [64]: %timeit jn(x[:, None], a)
846 µs ± 2.91 µs per loop (mean ± std. dev. of 7 runs, 1000 loops each)

In [65]: (jn(x[:, None], a) == jvv(x, a)).all()
Out[65]: True
```

That aside, usage questions are better suited for the scipy-user mailing list or the scipy community slack workspace or  Stackoverflow --- pretty much anywhere which is not the issue tracker :-).

**Comment by dschmitz89** [comment]
You can compute all orders and points at once in a vectorized call:

```python
import numpy as np
from scipy.special import jn

x = np.arange(1e-3, 200, 1e-3)
j = np.arange(51, dtype=int).reshape(j.shape[0], -1)

jn(j, x)
``` 

This still takes 44 sec. on my machine but it far better than the reported 10 minutes.

**Comment by LemonBoy** [comment]
> That aside, usage questions are better suited for the scipy-user mailing list or the scipy community slack workspace or Stackoverflow --- pretty much anywhere which is not the issue tracker :-).

Well I opened a ticket because there's _slow_ and there's **slow**, I wanted to get some ideas about where the slowdown may come from (and possibly solve it :).

> This still takes 44 sec. on my machine but it far better than the reported 10 minutes.

With tens of seconds I meant something around 50/60s, which is still too much considering MATLAB only takes 5s.

I'm having some nice results by writing a thin cython wrapper around Cephes' `yn` and beside some small (haven't measured that yet) numerical discrepancies the runtime is around ~2s as expected... So I believe there's no reason for scipy's wrappers to be so slow, that's why I initially assumed the vectorization wrapper to be the culprit here.

**Comment by ev-br** [comment]
Well, ufunc machinery does add some overhead (no, it's not `np.vectorize`, by far), but it is typically only significant for scalar arguments or very small arrays. 

My little experiment above with going through cython_special show essentially the same timings for a single ufunc call (In [64]) and an explicit cython loop with no OpenMP. Which is consistent with my previous experience: ufunc overhead should be negligible at large array sizes. If you observe significant perf difference between `cython_special` wrappers and your manual wrappers, it'd be great to see where the problem is.

**Comment by LemonBoy** [comment]
Ok, I spent some more time playing around with Cython and got the following results:
```
CEPHES:    s/loop 2.0529773000162095
AMOS:      s/loop 32.802241533334985
AMOS(vec): s/loop 1.0743747000039245
SCIPY:     s/loop 49.487069599990114
```

Where `AMOS` is a Cython wrapper around [this AMOS port to C](https://github.com/jpcima/zbessel) (because I don't have a Fortran compiler on this laptop) and `AMOS(vec)` is the same but exploits the API's ability to compute orders `fnu...fnu+N` at once for the given point.

I've checked the agreement between AMOS and SciPy and, as expected, they match very well. Cephes and AMOS start disagreeing at low-ish orders (as explained in #509) with such narrowly-spaced sample points.

**Comment by ev-br** [comment]
I think it would be very helpful to see the full code of this benchmark of yours. There does seem to be something to fix, could we tempt you to take a closer look?
I'll also ping the scipy.special resident  @steppi .

**Comment by LemonBoy** [comment]
Sure thing, I've uploaded the test code [here](https://github.com/LemonBoy/fast-bessel). 

> There does seem to be something to fix, could we tempt you to take a closer look?

I'll certainly try to find some time to contribute something back to SciPy :)

**Comment by steppi** [comment]
I've been a bit busy lately, but will look into look into this within 2-3 weeks.

**Comment by LemonBoy** [comment]
Gentlest ping?

**Comment by ilayn** [comment]
Hi @LemonBoy I don't know if you saw it but there are some activities finally happening https://github.com/scipy/scipy/issues/19404 so still definitely interested

**Comment by ilayn** [comment]
After #19587 I've replicated the results with the new C code. The `ufunc` machinery + our wrappers are doing something that causes things to go 30-40 seconds. But both single point C calls and the cythonization shows that things speed up dramatically invalidating the extra wait times. We might have an issue here indeed.

**Comment by JaRoSchm** [comment]
Would it be possible to expose the vectorization functionality for consecutive orders of Bessel functions implemented in amos (and its new C port) through scipy? I have a similar usage scenario like @LemonBoy mentioned above but additionally including `hankel1`.

I could imagine two ways of doing this:
1. Checking for each call to `jv`, `hankel1`, etc. if the order `v` is an array of consecutive integers (using `np.ediff1d`?) and if that's the case using this functionality from amos automatically.
2. Exposing this functionality through `cython_special` such that we can write our own Python wrappers around this.

At the moment in our project we would have to use an additional C port of amos to make use of this functionality although it is already included (but not public) in scipy.

I'm happy to help with the implementation of this if there is some consent.

**Comment by ilayn** [comment]
There is some fat that we need to remove in the ufunc machinery. Unfortunately I am not a `special` resident hence I don't have all the legacy information however as made obvious in this issue too, there is some hope that we can make a leaner linkage to ufunc machinery hopefully that would make the exposure more flexible. 

Having said that I think if `special` residents agree, then number 2 is the least troubling option. It seems somehow that Cython sitting in between ufunc engine and C code is slowing us down. This was necessary because of integrating Fortran is, not fun, to be polite. As we now removed that hurdle, I think we should investigate a Cythonless ufunc machinery and then probably, your request and potentially many similar other requests would already be at the acceptable execution speed by even custom Python wrappers.

**Comment by JaRoSchm** [comment]
While I'm definitely looking forward to these improvements too, I think their performance improvement is unrelated to my proposal. While removing the "fat" will (according to the benchmarks above) at most give a factor 2, "vectorizing" the orders can seemingly give an improvement on the order of number of orders. If however developer time or implementation hurdle in the current machinery is your main concern, I definitely understand this prioritization.

If its easier to keep track of my proposal in a separate issue in the long run, I can do open one.

**Comment by ilayn** [comment]
Oh then probably it completely flew above my head. Apologies for that. I thought vectorizing things would get this functionality out of the box. It seems like I confused the use case.

**Comment by JaRoSchm** [comment]
Oh so just to be clear to everyone what I mean by "vectorization": As Bessel functions are partly computed in a recursive way, one can keep the intermediate Bessel functions of different order which therefore don't have to be computed multiple times. As far as I can see by a quick glance, this is explained by Amos in one of his [papers](https://www.osti.gov/biblio/6148453) in section 9. This is implemented in the amos library and your C port, see e.g. https://github.com/scipy/scipy/blob/4cfd7518526f0c2cd3b1f2cdaab7393d986b72d5/scipy/special/_amos.c#L1170

In `amos_wrappers.c` this is simply set to 1 (i.e. not used): https://github.com/scipy/scipy/blob/4cfd7518526f0c2cd3b1f2cdaab7393d986b72d5/scipy/special/amos_wrappers.c#L764

I propose to expose this functionality through `amos_wrappers.c` (minimal implementation, however likely to substantially change soon), `cython_special` (more stable but one has to know about it) and/or the usual Python interface (automatically used? New keyword argument like in amos?).

**Comment by steppi** [comment]
I think exposing computation of Bessel functions to multiple orders would be a good feature to add. At the Python level, my preference would be to add new functions, rather than adding checks for consecutive integers in the order parameter. This would be similar to how we have both [lpmv](https://docs.scipy.org/doc/scipy/reference/generated/scipy.special.lpmv.html#scipy.special.lpmv) and [lpmn](https://docs.scipy.org/doc/scipy/reference/generated/scipy.special.lpmn.html#scipy.special.lpmn) for associated Legendre functions.

 In [cython_special](https://docs.scipy.org/doc/scipy/reference/special.cython_special.html), we could add these to the bucket of custom functions which are not ufuncs.

**Comment by JaRoSchm** [comment]
Thank you! This sounds like a good proposal. I will try to propose an implementation in the coming days or weeks. Probably I will start with a POC for just one of the Bessel functions.

**Comment by JaRoSchm** [comment]
After thinking a bit more about the implementation, I think this is a bit beyond my level of C and possibly Cython. The signature of these functions could be something like `jv_seq(v: float, n: int, z: complex)`, which therefore would not be ufuncs. `v` would be the initial order and `n` the number of orders in the sequence. If `z` would be an ndarray, I'm not sure which output shape should be used ((n, z.shape)?). Due to that I thought about skipping these `amos_wrapper.c` things which are only needed to generate ufuncs. Instead a Cython wrapper would be enough (and a Python function in `_basic.py`), but this would imply to translate all the error flag handling etc. from `amos_wrapper.c` (see e.g. `cbesj_wrap`). I'm not sure if it is wanted from your side to have all these things at two different places. Additionally, I do not understand all the things going on there and could therefore only translate them line by line.

Therefore, I think its its more helpful to you to leave this implementation to someone with a deeper understanding of `scipy.special`. If there is a clear way forward to follow, I'm happy to help translating the remaining Bessel functions.

Here is the small Cython wrapper I came up with:
```python
import numpy as np
cimport numpy as np

cdef extern from "_amos.h" nogil:
    int amos_besj(double complex z, double fnu, int kode, int n, double complex *cy,
                  int *ierr)

def cbesj_wrap_seq(double v, int n, np.npy_cdouble z):
    cdef double complex cy
    cdef int ierr = 0
    cdef double complex z_c = np.npy_creal(z) + np.npy_cimag(z)*1j

    amos_besj(z_c, v, 1, n, &cy, &ierr)

    return np.asarray(cy)
```

**Comment by JaRoSchm** [comment]
> After #19587 I've replicated the results with the new C code. The `ufunc` machinery + our wrappers are doing something that causes things to go 30-40 seconds. But both single point C calls and the cythonization shows that things speed up dramatically invalidating the extra wait times. We might have an issue here indeed.

As far as I can see, there were some recent changes to these wrappers. Therefore, I re-run the benchmark from https://github.com/scipy/scipy/issues/19079#issuecomment-1682143323.

SciPy 1.12.0:
```
CEPHES:    s/loop 0.09825446900019112
AMOS:      s/loop 0.4903019743336093
AMOS(vec): s/loop 0.10132556333337561
SCIPY:     s/loop 0.6610271896667351
```

main:
```
CEPHES:    s/loop 0.09950305633416671
AMOS:      s/loop 0.49242116433379124
AMOS(vec): s/loop 0.10114642533274794
SCIPY:     s/loop 0.5338746980002421
```
Some improvement can indeed be observed.

The benchmark from the original post (using `number=1`) gives 22.1s for SciPy 1.12.0. and 25.8s on main. 
Are there additional attempts to address this issue?


Concerning my idea to expose the "vectorization" functionality from amos (https://github.com/scipy/scipy/issues/19079#issuecomment-1917449792), I could open a separate issue, although this could help to address the original issue too. Maybe this is what is used internally in Matlab?
To explain a bit why I failed to implement this myself: This doesn't work like the other amos wrappers (scalar kernel which is vectorized) but is already vectorized on the level of amos. Therefore, some other kind of wrapper would be needed. Maybe something along the lines of https://github.com/scipy/scipy/pull/20320 is needed for that?

**Comment by izaid** [comment]
@JaRoSchm Our intention is to do this kind of vectorisation for all relevant special functions, of which Bessel is one. Yes, https://github.com/scipy/scipy/pull/20320 is the relevant machinery for that. I think we should be able to address your issue quite soon.

**Comment by JaRoSchm** [comment]
Thank you for working on these things! I‘m really looking forward to the results of your work on special.

**Comment by izaid** [comment]
> Thank you for working on these things! I‘m really looking forward to the results of your work on special.

I have literally wanted these things for years too!
